[RFC] Fix msg_puts_printf() on windows

erw7 opened the pull request:
This PR fixes the following issues:

- `nvim --version`: NVIM v0.4.0-657-ga0d723db5
- Vim (version: ) behaves differently? Yes(8.1.1397)
- Operating system/version: Microsoft Windows [Version 6.1.7601]
- Terminal name/version:
- `$TERM`:

### Steps to reproduce using `nvim -u NORC`

```
set LANG=ja_JP
nvim -u NORC --cmd "echomsg 'Hello'"
```

### Actual behaviour
Messages are displayed in garbled.

```
Hello
邯壹￠繧九↓縺ｯENTER繧呈款縺吶°繧ｳ繝槭Φ繝峨ｒ蜈･蜉帙＠縺ｦ縺上□縺輔＞
```

### Expected behaviour
The message is displayed normally.

```
Hello
続けるにはENTERを押すかコマンドを入力してください
```

erw7 committed:
Fix garbled problem with msg_puts_printf on Windows

erw7 committed:
Add test for #7967

bfredl commented on src/nvim/vim.h:
```diff
@@ -287,9 +291,34 @@ enum { FOLD_TEXT_LEN = 51 };  //!< buffer size for get_foldtext()
 // is declared.
-#define mch_errmsg(str)        fprintf(stderr, "%s", (str))
+#ifdef WIN32
+# define mch_errmsg(str) \
```
Why not write these as functions?

erw7 committed:
Change mch_errmsg and mch_msg from macro to function

justinmk commented:
Is this needed after https://github.com/neovim/neovim/pull/9829? Would much rather remove `USE_MCH_ERRMSG` entirely and go forward with https://github.com/neovim/neovim/pull/9829. If it's blocked by https://github.com/neovim/neovim/pull/9338 I will try to unblock it .

justinmk commented on test/functional/core/startup_spec.lua:
```diff
@@ -219,6 +219,29 @@ describe('startup', function()
     eq(2, eval('1+1'))
   end)
+
+  it('message display ends with enter #7967', function()
```
#7967 is fixed on master, right? So we can add this test already.

erw7 replied:
Yes, #7967 has been fixed by # 9884.

erw7 committed:
Remove USE_MCH_ERRMSG

USE_MCH_ERRMSG has never been defined, so the dead code has been removed.

erw7 commented:
> Is this needed after #9829?

Even after #9829 `nvim -V1 -Es` needs this. For #9338 I created #10109.

justinmk commented:
> Even after #9829 `nvim -V1 -Es` needs this. 

If it's not much trouble, can you add a test for that?

justinmk commented on src/nvim/vim.h:
```diff
@@ -287,9 +287,11 @@ enum { FOLD_TEXT_LEN = 51 };  //!< buffer size for get_foldtext()
+# define mch_errmsg(str)        fprintf(stderr, "%s", (str))
+# define mch_msg(str)           printf("%s", (str))
+#endif
 #define display_errors()       fflush(stderr)
```
`display_errors` could be removed completely. It is only used in `do_shell()`, left over from Vim. `do_shell()` in Nvim uses the Nvim UI, not the tty directly.

erw7 committed:
Remove display_erros()

do_shell() in Nvim uses the Nvim UI, not the tty directly, so
display_errors() is not necessary anymore.

erw7 commented:
>> Even after #9829 nvim -V1 -Es needs this.

> If it's not much trouble, can you add a test for that?

@justinmk  The test environment does not use message catalogs, because `LOCALE_INSTALL_DIR` is passed to` bindtextdomain()`. This is also true for appimage and Windows  artifacts. We need to solve the problem before creating a test.

I'm thinking of changing to `exepath/../ share/locale` instead of `LOCALE_INSTALL_DIR`, like `os_setenv_append_path()`. Do you have any other good ideas? Do I need to create a new issue on this matter?

justinmk commented:
> I'm thinking of changing to `exepath/../ share/locale` instead of `LOCALE_INSTALL_DIR`, like `os_setenv_append_path()`. 

:+1: that seems completely fine.

erw7 committed:
Fix problems with message catalog directory

- In appimage, the message catalog is not used because there is no
  message catalog in LOCALE_INSTALL_DIR. Therefore, change to
  exepath/../share/locale instead of LOCALE_INSTALL_DIR.
- The old vim style($runtime/lang) is no longer used. Thus all relevant
  code is removed.

erw7 committed:
Add msg_puts_printf() test for multibyte characters

justinmk commented:
```
../src/nvim/main.c:724:61: error: implicit conversion changes signedness: 'long' to 'unsigned long' [-Werror,-Wsign-conversion]
  xstrlcpy(tail, "share/locale", sizeof(localepath) - (tail - localepath));
                                                    ~  ~~~~~^~~~~~~~~~~~
```

justinmk commented on src/nvim/main.c:
```diff
@@ -738,22 +738,17 @@ static void init_locale(void)
+  *tail = NUL;
+  tail = (char *)path_tail((char_u *)exepath);
+  xstrlcpy(tail, "share/locale", sizeof(localepath) - (tail - localepath));
+  bindtextdomain(PROJECT_NAME, localepath);
```
does not regress 1a15cf84c29c984d600f438122edc70be2ac91a1 as long as we assume `@CMAKE_INSTALL_FULL_LOCALEDIR@` would always be in this form.

I wonder if this conforms with BSD conventions.

erw7 committed:
Change to use VV_PROGPATH instead os_exepath()

erw7 committed:
Change to not test msg_puts_pirntf() in unix CI
